### PR TITLE
fix(CORE): C++ 11 rule of 3 compiant constructors

### PR DESCRIPTION
**Pull request by mishaparem**
C++ 11 Rule of Three:
If a class requires a user-defined destructor, a user-defined copy
constructor, or a user-defined copy assignment operator, it almost
certainly requires all three.
https://en.cppreference.com/w/cpp/language/rule_of_three

Constructors added in commit remove warning during build, so as to
build successfully without warnings.

https://github.com/azerothcore/azerothcore-wotlk/issues/3021

## CHANGES PROPOSED:
-  Explicitly defaults destructor, copy constructor, or copy assignment operator, if one has been defined by user.

## ISSUES ADDRESSED:
- Closes 

Closes https://github.com/azerothcore/azerothcore-wotlk/issues/3021

## TESTS PERFORMED:
- Build performed, confirmed that no constructor warnings occur.
- Test performed on Linux (OpenSuse Tumbleweed)

## HOW TO TEST THE CHANGES:

- Build, view warnings. If no constructor warnings, test passed.
- Changes should not affect gameplay or core functionality.

## KNOWN ISSUES AND TODO LIST:

- [x] No foreseeable issues with these changes.

## Target branch(es):
- [x] Master commit 2bb868164e545d38da5c96a3ce683a5fcb3b72a6

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: deps/g3dlite/include/G3D/Vector3.h [hunk 1]
```diff
@@ -116,6 +116,11 @@ class Vector3 {
 
     // assignment and comparison
     Vector3& __fastcall operator= (const Vector3& rkVector);
+    /* requried as of C++ 11 */
+    #if __cplusplus >= 201103L
```
Comment (Viste) [inline deps/g3dlite/include/G3D/Vector3.h:120]: its not needed c++17 enabled

File: src/common/Packets/ByteBuffer.h [hunk 2]
```diff
@@ -72,6 +72,12 @@ class ByteBuffer
             _storage(buf._storage)
         {
         }
+        /* requried as of C++ 11 */
+        #if __cplusplus >= 201103L
```
Comment (Viste) [inline src/common/Packets/ByteBuffer.h:76]: its not needed c++17 enabled

File: src/common/Packets/WorldPacket.h [hunk 3]
```diff
@@ -22,6 +22,12 @@ class WorldPacket : public ByteBuffer
         WorldPacket(const WorldPacket &packet)              : ByteBuffer(packet), m_opcode(packet.m_opcode)
         {
         }
+        /* requried as of C++ 11 */
+        #if __cplusplus >= 201103L
```
Comment (Viste) [inline src/common/Packets/WorldPacket.h:26]: its not needed c++17 enabled

File: src/server/game/Entities/Object/Object.h [hunk 4]
```diff
@@ -376,6 +376,12 @@ struct Position
         : m_positionX(x), m_positionY(y), m_positionZ(z), m_orientation(NormalizeOrientation(o)) { }
 
     Position(Position const& loc) { Relocate(loc); }
+    /* requried as of C++ 11 */
+    #if __cplusplus >= 201103L
```
Comment (Viste) [inline src/server/game/Entities/Object/Object.h:380]: its not needed c++17 enabled

File: src/common/Utilities/Util.h [hunk 5]
```diff
@@ -449,6 +449,11 @@ class flag96
         part[2] = right.part[2];
         return *this;
     }
+    /* requried as of C++ 11 */
+    #if __cplusplus >= 201103L
```
Comment (Viste) [inline src/common/Utilities/Util.h:453]: its not needed c++17 enabled

File: src/server/game/Entities/Object/Object.h [hunk 6]
```diff
@@ -639,7 +645,13 @@ class WorldLocation : public Position
     public:
         explicit WorldLocation(uint32 _mapid = MAPID_INVALID, float _x = 0, float _y = 0, float _z = 0, float _o = 0)
             : m_mapId(_mapid) { Relocate(_x, _y, _z, _o); }
-        WorldLocation(const WorldLocation &loc) { WorldRelocate(loc); }
+        WorldLocation(const WorldLocation &loc) : Position () { WorldRelocate(loc); }
+        /* requried as of C++ 11 */
+        #if __cplusplus >= 201103L
```
Comment (Viste) [inline src/server/game/Entities/Object/Object.h:650]: its not needed c++17 enabled

File: src/server/game/Entities/Unit/Unit.h [hunk 7]
```diff
@@ -1314,6 +1314,12 @@ class MMapTargetData
         _posOwner.Relocate(c._posOwner);
         _posTarget.Relocate(c._posTarget);
     }
+    /* requried as of C++ 11 */
+    #if __cplusplus >= 201103L
```
Comment (Viste) [inline src/server/game/Entities/Unit/Unit.h:1318]: its not needed c++17 enabled

File: src/server/game/Movement/Spline/MoveSplineFlag.h [hunk 8]
```diff
@@ -68,6 +68,12 @@ namespace Movement
         MoveSplineFlag() { raw() = 0; }
         MoveSplineFlag(uint32 f) { raw() = f; }
         MoveSplineFlag(const MoveSplineFlag& f) { raw() = f.raw(); }
+        /* requried as of C++ 11 */
+        #if __cplusplus >= 201103L
```
Comment (Viste) [inline src/server/game/Movement/Spline/MoveSplineFlag.h:72]: its not needed c++17 enabled

File: deps/g3dlite/include/G3D/Vector3.h (hunk 1 above)
Comment (mishaparem) [reply deps/g3dlite/include/G3D/Vector3.h:120]: problem persists with -DCMAKE_CXX_FLAGS="-Werror -std=c++17"

Comment (Yehonal) [reply deps/g3dlite/include/G3D/Vector3.h:120]: https://en.cppreference.com/w/cpp/language/default_constructor

image

@Viste

**Comment by FrancescoBorzi** [comment]
did anyone test this in game?

**Comment by Yehonal** [comment]
> did anyone test this in game?

it shouldn't be needed but, eventually, we could just check that you can move your character in game without any crash

**Comment by FrancescoBorzi** [comment]
yes, just a quick check in game please

**Comment by gengarshadowball** [comment]
Tested, everything looks good 👍

**Comment by Kitzunu** [comment]
Ty for the PR!

**Pull request merged**